### ISSUE TITLE
Suggestion: Check argument types for `validate` function

### ISSUE THREAD

**Issue by mtalexan** [python-jsonschema/jsonschema#1106]
The `validate` function is the primary entrypoint for (possibly ignorant) users.  It requires the `instance` and `schema` arguments to be post-parsed Python data structures corresponding to the instance and schema, but an ignorant reader could easily assume they're supposed to be strings of the content or strings for file names instead.  Type checking in `validate` could catch this error early.

It only makes sense to operate with a schema that is a single JSON object, which always post-parses into a `dict`.  So it should be easy to just verify that the `schema` argument is a `dict`, which is how the `json` library will encode an object.  

Since the `json` library can't encode `json.loads(''' "key": "value" ''')`, and that's not valid JSON in the first place, there are really only two possible JSON structure types for the `instance` argument: array or object.  These correspond to the Python `list` and `dict` types respectively, so the `instance` can be checked to see if it's a `list` or `dict` type.

The internal details of the `schema` `dict` can't be defined, nor the internal details of the `instance` `dict|list`, but at least the top level type can be checked pretty trivially.  A simple implementation might be:
```
if not isinstance(instance, (list, dict)):
    raise TypeError("'instance' argument isn't a Python dict or list corresponding to a JSON object or array")
if not isinstance(schema, dict):
    raise TypeError("'schema' argument isn't a Python dict corresponding to a JSON object")
```
---

TL;DR  
Add checks to `validate` function to verify `isinstance(instance, (list, dict))` and `isinstance(instance, dict)`, or raise a clear `TypeError`.

---

I (stupidly) accidentally passed a string of the schema to a `validate` call as the `schema`.  Since the first thing `validate` does is call `validator_for` on the `schema`, and a non-empty string `is True` (from `if schema is True or schema is False or "$schema" not in schema:`), this resulted in a `TypeError` during the lookup of the `schema["$schema"]` value. This led me (also stupidly) to believe my schema's `$schema` value wasn't correct for quite a long time.  

If the type error had been caught earlier by the `validate` with a clear error message it would have been much easier to realize my actual error.  And (I believe) the solution is backwards compatible with all supported Python versions.

**Comment by Julian** [comment]
I'll leave this open to consider it at least slightly, though immediately my reaction is "this is generally not the way things are done conventionally in Python", in general it's simply documented what type things expect, though...

> It only makes sense to operate with a schema that is a single JSON object, which always post-parses into a dict

(Leaving aside that it's `Mapping` not `dict` that's expected for object schemas), this is only true for old versions of JSON Schema, in newer ones they can also be `bool`... and in theory for an arbitrary version of JSON Schema they can actually be any object, since `jsonschema` supports extending drafts with additional types (via `jsonschema.validators.extend`)! So it's not so simple.

As I say I'm happy to leave this open and think for more than the 30 seconds to see if there's some improvement I *am* comfortable with, and regardless, thanks for filing the request.

**Comment by Julian** [comment]
I'm going to close this for the reasoning above, but again still appreciate the idea.
